src/js/statements.rs

```diff
@@ -0,0 +1,145 @@
+    }
+}
+
+fn mod_fun<'a>(
+    public: &bool,
+    name: &'a str,
+    args: &'a [TypedArg],
+    body: &'a TypedExpr,
+    _module: &'a [String],
+    _return_type: &'a Arc<Type>,
+    _line_numbers: &'a LineNumbers,
+) -> Result<Document<'a>> {
+    let env = Env {
+        return_last: &true,
+        semicolon: &true,
+    };
+    let body = expr(body, &env)?;
+    Ok(if &true == public { "export " } else { "" }
+        .to_doc()
+        .append("function ")
+        .append(Document::String(name.to_string()))
+        .append(fun_args(args))
+        .append(" {")
+        .append(line().append(body).nest(INDENT).group())
+        .append(line())
+        .append("}"))
+}
+
+fn fun_args<'a>(args: &'a [TypedArg]) -> Document<'a> {
+    wrap_args(args.iter().map(|a| match &a.names {
+        ArgNames::Discard { .. } | ArgNames::LabelledDiscard { .. } => "_".to_doc(),
+        ArgNames::Named { name } | ArgNames::NamedLabelled { name, .. } => {
+            Document::String(name.to_string())
+        }
+    }))
+}
+
+fn wrap_args<'a, I>(args: I) -> Document<'a>
+where
+    I: Iterator<Item = Document<'a>>,
+{
+    break_("", "")
+        .append(concat(Itertools::intersperse(args, break_(",", ", "))))
+        .nest(INDENT)
+        .append(break_("", ""))
+        .surround("(", ")")
+        .group()
+}
+
+fn expr<'a>(expression: &'a TypedExpr, env: &Env<'a>) -> Result<Document<'a>> {
+    let rendered = match expression {
+        TypedExpr::String { value, .. } => Ok(string(value)),
+
+        TypedExpr::Int { .. } => unsupported("Integer values "),
+        TypedExpr::Float { .. } => unsupported("Float values"),
+
+        TypedExpr::List { .. } => unsupported("List "),
+
+        TypedExpr::Tuple { .. } => unsupported("Tuple "),
+        TypedExpr::TupleIndex { .. } => unsupported("Tuple "),
+
+        TypedExpr::Case { .. } => unsupported("Case "),
+
+        TypedExpr::Call { .. } => unsupported("Function "),
+        TypedExpr::Fn { .. } => unsupported("Function "),
+
+        TypedExpr::RecordAccess { .. } => unsupported("Custom Record "),
+        TypedExpr::RecordUpdate { .. } => unsupported("Function "),
+
+        TypedExpr::Var { .. } => unsupported("Bindings "),
+        TypedExpr::Seq { .. } => unsupported("Bindings "),
+        TypedExpr::Assignment { .. } => unsupported("Bindings "),
+
+        TypedExpr::BinOp { .. } => unsupported("Binary operation "),
+
+        TypedExpr::Todo { .. } => unsupported("todo keyword "),
+
+        TypedExpr::BitString { .. } => unsupported("Bitstring "),
+
+        TypedExpr::Pipe { .. } => unsupported("Pipe "),
+
+        TypedExpr::ModuleSelect { .. } => unsupported("Module function call "),
+    }?;
+    Ok(match expression {
+        TypedExpr::Seq { .. } | TypedExpr::Assignment { .. } => rendered,
+        _ => match env.return_last {
+            true => "return ",
+            _ => "",
+        }
+        .to_doc()
+        .append(rendered)
+        .append(match env.semicolon {
```

Comment:
I think rather than considering `;` part of the expression I would prefer to have it be part of the JS generation for `Seq`, `Assignment` and `Fn` AST nodes. That way we don't need to mutate this state and check each time.

Reply:
I tried that, it might be possible, but I struggled to make it work.
Because JS doesn't do blocks properly I rendered Gleam blocks to comma expression in JS, at which point you have let expressions that end in a comma.

Reply:
Cool I'll see if I can work something out.